### DISCUSSION TITLE
GatedGraphConv implementation on link prediction

### DISCUSSION THREAD

**Discussion by haoyuan-l** [pyg-team/pytorch_geometric#6093] [Q&A]
![image](https://user-images.githubusercontent.com/91633635/204665124-8b143c81-bba7-4d0f-b6fe-cdc0f1d98ab5.png)
Hi, I tried to implement the above architecture from the paper [https://arxiv.org/abs/1511.05493](Gated Graph Sequence Neural Networks). But I'm confused with the PyG `GatedGraphConv` operator, is this operator works as a single GG-NN in the network?(e.g. like _Fo_&_Fx_ in the above figure)

I didn't see implementations of the node annotation output model for predicting **_X_**(k+1) from **_H_**(k,T) in the source code. This is an essential step for the next node prediction. If I set the `out_channels` equal to the input size, how can I identify the next link node?

Input and output for the `GatedGraphConv` operator. (each node has 3 features)
![image](https://user-images.githubusercontent.com/91633635/204667008-c15ad2eb-c88c-465b-9a19-59c6b36ee7e4.png)

**Comment by rusty1s** [comment]
Yes, our `GatedGraphConv` model only implements the "Propagation Model" as given in Section 3.2. You can use these layers to construct the GGS-NN model. Let me know if that resolves your concerns.

**Reply by haoyuan-l** [reply]
@rusty1s Thank you for your reply. I still don't understand the output of the operator `GatedGraphConv`. Is this a hidden state for the next node or node score? If it is a hidden state for the next step, how can it be used for node selection (i.e., identify the node)?
![image](https://user-images.githubusercontent.com/91633635/204806858-970ec97e-f0a0-4557-a0c2-766374d0e33f.png)
In the above figure, how can the output of `GatedGraphConv` feed into the output model like **ov** = _**g**_(**h**v(T) , **x**v) ?

**Reply by rusty1s** [reply]
`GatedGraphConv` will just return new node embeddings, e.g., it starts from an embedding `h_v^(1)` and produces `h_v^(T)`. For the node selection, take a look at the `aggr.AttentionalAggregation` operator.

**Comment by haoyuan-l** [comment]
<img width="457" alt="image" src="https://user-images.githubusercontent.com/91633635/205948077-7db05309-eb47-44a1-a58e-94e51e5e9665.png">
Hi @rusty1s , I'm kind of confused with this part of the source code for `GatedGraphConv`. Why do you use a for loop here but only output once? I noticed that the iteration times of the for-loop equal to the sequence length, does the purpose of the loop here is to do the message propagation based on the prediction length?

**Reply by rusty1s** [reply]
This model will compute h_v^(T), that is, it iteratively computes embeddings from h_v^(1) to h_v^(T) to derive at the final node representation. This is basically Eq. (1) to Eq. (6) applied multiple times to derive at Eq. (7). Please correct me if I am wrong.

**Reply by haoyuan-l** [reply]
So `GatedGraphConv` only outputs the final node representation for the sequence? What if I want to output each step for the prediction to get the whole sequence? (i.e. the next node, because my purpose is to do path prediction) Should I override the forward function to output each iteration result during the for-loop?

**Reply by rusty1s** [reply]
There is no sequence prediction integrated here, this is just the the propagation model described in Section 3. For sequence prediction, you need to follow Section 4, e.g., a `GatedGraphConv` propagation model is used to predict the next token in the sequence. Hope that clarifies your concerns.

**Reply by haoyuan-l** [reply]
Thank you for your reply. Should I remove the for-loop in the forward function to use `GatedGraphConv` for the next token prediction? (In the path-finding problem, because there is only one timestep between each connected node in the graph, right?)

**Reply by rusty1s** [reply]
I think you would need to keep it. The proposed model is designed to perform multiple steps of propagation.

**Reply by haoyuan-l** [reply]
Thank you, one more question, should I add an `aggr.AttentionalAggregation` layer after `GatedGraphConv` to do the sequence prediction?

**Reply by rusty1s** [reply]
Yes, I think so.
